winml/lib/Api.Ort/OnnxruntimeEngine.cpp

```diff
@@ -0,0 +1,1132 @@
+  return S_OK;
+}
+
+HRESULT OnnxruntimeEngine::Sync() {
+  auto winml_adapter_api = engine_factory_->UseWinmlAdapterApi();
+
+  OrtExecutionProvider* ort_provider;
+  winml_adapter_api->SessionGetExecutionProvider(session_.get(), 0, &ort_provider);
+
+  winml_adapter_api->ExecutionProviderSync(ort_provider);
+  return S_OK;
+}
+
+OrtSession* OnnxruntimeEngine::UseOrtSession() {
+  return session_.get();
+}
+
+const OrtApi* OnnxruntimeEngine::UseOrtApi() {
+  return engine_factory_->UseOrtApi();
+}
+
+OnnxruntimeEngineFactory* OnnxruntimeEngine::GetEngineFactory() {
+  return engine_factory_.Get();
+}
+
+HRESULT OnnxruntimeEngine::CreateTensorValue(const int64_t* shape, size_t count, winml::TensorKind kind, _Out_ IValue** out) {
+  auto ort_api = engine_factory_->UseOrtApi();
+  auto winml_adapter_api = engine_factory_->UseWinmlAdapterApi();
+
+  OrtExecutionProvider* ort_provider;
+  winml_adapter_api->SessionGetExecutionProvider(session_.get(), 0, &ort_provider);
+
+  OrtAllocator* ort_allocator;
+  winml_adapter_api->GetProviderAllocator(ort_provider, &ort_allocator);
+  auto unique_allocator = UniqueOrtAllocator(ort_allocator, winml_adapter_api->FreeProviderAllocator);  // the release here should probably not return anything
+
+  OrtValue* ort_value;
+  ort_api->CreateTensorAsOrtValue(unique_allocator.get(), shape, count, ONNXTensorElementDataTypeFromTensorKind(kind), &ort_value);
+  auto unique_value = UniqueOrtValue(ort_value, ort_api->ReleaseValue);
+  RETURN_IF_FAILED(Microsoft::WRL::MakeAndInitialize<OnnxruntimeValue>(out, this, std::move(unique_value), std::move(unique_allocator)));
+  return S_OK;
+}
+
+using DmlAllocatorResource = std::unique_ptr<void, void (*)(void*)>;
+class DmlAllocatorWrapper : public Microsoft::WRL::RuntimeClass<
+                                Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::ClassicCom>,
+                                IUnknown> {
+ public:
+  DmlAllocatorWrapper() : dml_resource_(nullptr, nullptr) {}
+
+  HRESULT RuntimeClassInitialize(DmlAllocatorResource&& dml_resource) {
+    dml_resource_ = std::move(dml_resource);
+    return S_OK;
+  }
+
+ private:
+  DmlAllocatorResource dml_resource_;
+};
+
+HRESULT OnnxruntimeEngine::CreateTensorValueFromExternalD3DResource(ID3D12Resource* d3d_resource, const int64_t* shape, size_t count, winml::TensorKind kind, _Out_ IValue** out) {
+  auto ort_api = engine_factory_->UseOrtApi();
+  auto winml_adapter_api = engine_factory_->UseWinmlAdapterApi();
+
+  OrtExecutionProvider* ort_provider;
+  winml_adapter_api->SessionGetExecutionProvider(session_.get(), 0, &ort_provider);
+
+  OrtMemoryInfo* dml_memory = nullptr;
+  winml_adapter_api->GetProviderMemoryInfo(ort_provider, &dml_memory);
+
+  void* dml_allocator_resource;
+  winml_adapter_api->DmlCreateGPUAllocationFromD3DResource(d3d_resource, &dml_allocator_resource);
+  auto unique_dml_allocator_resource =
```

Comment:
>unique_dml_allocator_resource [](start = 7, length = 29)

pass this into onnxruntime value?

Reply:
Cached on SetParameter.

---
In reply to: [370868996](https://github.com/microsoft/onnxruntime/pull/2891#discussion_r370868996) [](ancestors = 370868996)